File: typescript/nominal4.ts
abstract class Shape {
  abstract area(): number
}

class Rectangle extends Shape {
  constructor(private width: number, private height: number) {
    super()
  }
  area = () => this.width * this.height
}

class EquilateralTriangle extends Shape {
  constructor(private sideLen: number) {
    super()
  }
  area = () => Math.sqrt(3) / 4 * this.sideLen ** 2
}
export const roundArea = (shape: Shape) => Math.round(shape.area())


class SwimmingPool {
  area = () => 55.5
}

roundArea(new Rectangle(100, 50))
roundArea(new EquilateralTriangle(5))
// https://www.typescriptlang.org/docs/handbook/type-compatibility.html
roundArea(new SwimmingPool())

File: python/nominal4.py
from abc import ABCMeta, abstractmethod
from math import sqrt
from typing import Union


class Shape(metaclass=ABCMeta):
    @abstractmethod
    def area(self) -> Union[int, float]: ...


class Rectangle(Shape):
    def __init__(
        self,
        width: int,
        height: int,
    ):
        self.width = width
        self.height = height

    def area(self) -> int:
        return self.width * self.height


class EquilateralTriangle(Shape):
    def __init__(self, side_len: int):
        self.side_len = side_len

    def area(self) -> float:
        return sqrt(3) / 4 * self.side_len ** 2


class SwimmingPool:
    def area(self) -> float:
        return 50.5


def round_area(shape: Shape):
    return int(round(shape.area()))


round_area(Rectangle(100, 50))
round_area(EquilateralTriangle(5))
round_area(SwimmingPool())
